ADDED: Extended the section on working with tree-shaped data.

wouterbeek opened the pull request:
- Added example data.
- Clarified that the source format is abstracted away from (XML/JSON/other).
- Explained how list elements can be accessed by index.
- Explained how list elements can be iterated over.
- Explained the special meaning of `$index`, `$parent`, and `$root`.
- Added an example where `$parent` != `$root`.
- Explained special value `"__circular"`.

wouterbeek committed:
ADDED: Extended the section on working with tree-shaped data.

- Added example data.
- Clarified that the source format is abstracted away from (XML/JSON/other).
- Explained how list elements can be accessed by index.
- Explained how list elements can be iterated over.
- Explained the special meaning of `$index`, `$parent`, and `$root`.
- Added an example where `$parent` != `$root`.
- Explained special value `"__circular"`.

wouterbeek committed:
FIXED: Two small typo's.

dionenak commented on docs/ratt-transform/index.md:
````diff
@@ -675,84 +675,384 @@ app.use(
+}
+```
+
+Since the RATT record abstracts away the syntactic specifics of the source system format, the above example may be read from a JSON source (see [connecting JSON sources](#json)), or from another kind of source.  For example, the RATT record may be read from the following XML source, assuming the contents of the `<root>` tag are selected:
````
This link doesn't work for me.

wouterbeek replied:
I've added section for connecting JSON sources.

dionenak commented on docs/ratt-transform/index.md:
````diff
@@ -675,84 +675,384 @@ app.use(
+</root>
+```
+
+See [the section on connecting XML sources](#xml) for more information.
````
This link goes to github (in the same [page](https://github.com/TriplyDB/Documentation/blob/tree-shaped-data/docs/ratt-transform/index.md) but not to any specific section.

wouterbeek replied:
This section exists, but on a different page... fixed.

dionenak committed:
change 'labels' to 'countries' in tree-shapes

dionenak committed:
fix a sentence in tree shapes

dionenak committed:
fix a typo

dionenak commented on docs/ratt-transform/index.md:
```diff
@@ -675,84 +675,384 @@ app.use(
+
+RATT is able to access specific elements from such lists based on their *index* or position.  Following the standard practice in Computer Science, RATT refer to the first element in the list as having index 0.  The second element has index 1, etc.
+
+For the above example record, we can assert the first label of the country:
```
These examples are not about list elements, but objects in an array. It is important to mention that because later RATT works differently with objects that have arrays as values ( `mw.toLiteral.forEach()` and `mw.toIri.forEach()`) and objects with arrays of objects (`mw.forEach()`).

wouterbeek replied:
It is indeed good to think about the terminology here: JSON uses "array" for ordered lists. XML is not very explicit about how ordered lists are called, talking about the plural "elements" inside a parent "element".  I'm dropping my use of "sequence" here, because that term is neither used in JSON nor XML.  Let's go with "ordered list" that is more descriptive and less implementation-specific than "array" or "elements".

I agree that the user should be warned that `forEach` behaves differently on lists of objects than on lists of strings or numbers, but that warning should be part of the documentation for `forEach`.

wouterbeek replied:
Ah, `forEach` follows quickly in the next section I see...

wouterbeek replied:
`mw.toIri.forEach` and `mw.toLiteral.forEach` are not documented yet...

wouterbeek replied:
I have added a new section for iterating over lists of primitives.  The section for iterating over lists of objects refers to this new section.

I have made the terminology more consistent: "list" for this kind of data, "array" as the implementation in JSON, "elements" as the implementation in XML.

dionenak commented on docs/ratt-transform/index.md:
````diff
@@ -675,84 +675,384 @@ app.use(
+country:de rdfs:label 'Germany'@en.
+```
+
+### Iterating over lists
````
Iterating over arrays of objects.

wouterbeek replied:
I'm going to think about this a bit more.  I don't like having different documentation sections for this.

wouterbeek replied:
Renamed the section accordingly.

dionenak commented on docs/ratt-transform/index.md:
```diff
@@ -675,84 +675,384 @@ app.use(
+
+### Special keys when iterating over lists
+
+In the previous section we saw that `mw.forEach` allows us to iterate over the elements in an list.  The elements are themselves made available as RATT records.  The RATT records inside an `mw.forEach `function are smaller.  This allow the regular keys of the iterated-over elements to be accessed directly.
```
'over objects in an array' instead of 'elements in a list'.

wouterbeek replied:
Merged this section with the section called "Iterating over lists of objects".

dionenak commented on docs/ratt-transform/index.md:
```diff
@@ -675,84 +675,384 @@ app.use(
+
+Every RATT record inside an` mw.forEach` call contains the `"$root"` key.  The value of the root key provides a link to the full RATT record.  Because the `$root` key is part of the linked-to RATT record, it is not possible to print the value of the root key.  (This would result in infinite output.)  For this reason, the value of the `$root` key is printed as the special value `"__circular__"`.
+
+For the above examples, the parent record and root record are the same, but this is not always the case.  Specifically, the parent record and root record are different when `mw.forEach` calls are nested.
```
The example of nested forEach is complicated, I find it hard to follow. I would propose to use different data. If we use the above data : 
```
{
  "metadata": {
    "title": {
      "name": "Data about countries."
    }
  },
  "data": {
    "countries": [
      {
        "id": "nl",
        "name": "The Netherlands"
      },
      {
        "id": "de",
        "name": "Germany"
      }
    ]
  }
}
```
and iterate over 'data.countries', the `$parent `would be data but the `$root` will show the whole record including the sibling 'metadata'. Right?

wouterbeek replied:
I agree that nested `mw.forEach` + nested example data is more complicated.

But if we iterate over `'data.countries'` we get records like this (where `$parent`==`$root`):
```
{
  "id": "nl",
  "name": "The Netherlands",
  "$index": 0,
  "$parent": {
    "metadata": {
      "title": {
        "name": "Data about countries."
      }
    },
    "data": {
      "countries": [
        {
          "id": "nl",
          "name": "The Netherlands"
        },
        {
          "id": "de",
          "name": "Germany"
        }
      ]
    }
  },
  "$root": "__circular__"
}
```

I think the rule here is that `$parent` directly contains the first child in the specified sequence.  E.g. `$parent` directly contains `data` in this case.

wouterbeek replied:
Unfortunately, the more complex example is necessary in this case.  This is left as-is.

dionenak replied:
you're right! My mistake!

dionenak commented on docs/ratt-transform/index.md:
````diff
@@ -675,84 +675,384 @@ app.use(
+
+```
+[1] metadata.title.name
+[2] data.countries.name
````
the second path is not correct: `countries` has an array of objects, while `title` includes an object (which makes path [1]  correct). Thus,
[2] data.countries[x].name where x can be 0 or 1.

wouterbeek replied:
Ah, you're right.  That breaks my example :-(

I've changed this and refer to the next section where we explain what `[x]` does.

wouterbeek committed:
ADDED: Small 'stub' section on JSON sources (fixes internal link).

wouterbeek committed:
FIXED: XML should have HTML ID 'xml' (was: 'excel').

wouterbeek committed:
FIXED: Broken link for #xml.

wouterbeek committed:
DOC: Processed the feedback by Dionysia.

- Simplified the introduction of tree-shaped data (JSON and XML
examples).

- Made the data examples consistent ('value' → 'name').

- Fixed the example where index notation is necessary.

- Make explcit that mw.forEach only iterates over lists of *objects*.

- Tweaks to definitions of '$root' and '$parent'.

- New section "Iterating over lists of primitives".

wouterbeek commented:
@dionenak Thanks for the good feedback!  Can you check my latest changes?

TanjaRonzhina committed:
Updated index.md, corrected typos
